File: apps/token-proxy/token_proxy/limits.py
from __future__ import annotations

import json
import logging
from dataclasses import dataclass

from redis.asyncio import Redis
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

logger = logging.getLogger(__name__)

LIMIT_CACHE_PREFIX = "limit:"
LIMIT_CACHE_TTL = 60  # 60 seconds


@dataclass(frozen=True)
class LimitResult:
    allowed: bool
    warning: bool
    used: int
    limit: int
    tier: str


async def check_limits(
    customer_id: str,
    redis: Redis,
    db: AsyncSession,
) -> LimitResult:
    """Check token usage limits for a customer. Returns LimitResult."""
    cache_key = f"{LIMIT_CACHE_PREFIX}{customer_id}"

    # Try Redis cache
    cached = await redis.get(cache_key)
    if cached is not None:
        data = json.loads(cached)
        used, limit, tier = data["used"], data["limit"], data["tier"]
    else:
        # Cache miss — query DB
        result = await db.execute(
            text(
                "SELECT um.tokens_used, um.tokens_limit, s.tier "
                "FROM usage_monthly um "
                "JOIN subscriptions s ON s.customer_id = um.customer_id "
                "WHERE um.customer_id = :cid "
                "  AND um.period_start <= now() "
                "  AND um.period_end > now() "
                "  AND s.status = 'active'"
            ),
            {"cid": customer_id},
        )
        row = result.fetchone()

        if row is None:
            # No active subscription/usage record — block by default
            return LimitResult(allowed=False, warning=False, used=0, limit=0, tier="unknown")

        used, limit, tier = int(row[0]), int(row[1]), str(row[2])

        # Populate cache
        await redis.set(
            cache_key,
            json.dumps({"used": used, "limit": limit, "tier": tier}),
            ex=LIMIT_CACHE_TTL,
        )

    allowed = used < limit
    warning = used >= int(limit * 0.9)

    return LimitResult(allowed=allowed, warning=warning, used=used, limit=limit, tier=tier)

File: apps/token-proxy/src/limits.js
const LIMIT_CACHE_PREFIX = "limit:";
const LIMIT_CACHE_TTL = 60;

/**
 * Check token usage limits.
 * @param {string} customerId
 * @param {import("ioredis").default} redis
 * @param {import("pg").Pool} pg
 * @returns {Promise<{allowed: boolean, warning: boolean, used: number, limit: number, tier: string}>}
 */
export async function checkLimits(customerId, redis, pg) {
  const cacheKey = `${LIMIT_CACHE_PREFIX}${customerId}`;

  const cached = await redis.get(cacheKey);
  let used, limit, tier;

  if (cached !== null) {
    const data = JSON.parse(cached);
    used = data.used;
    limit = data.limit;
    tier = data.tier;
  } else {
    const { rows } = await pg.query(
      `SELECT um.tokens_used, um.tokens_limit, s.tier
       FROM usage_monthly um
       JOIN subscriptions s ON s.customer_id = um.customer_id
       WHERE um.customer_id = $1
         AND um.period_start <= now()
         AND um.period_end > now()
         AND s.status = 'active'`,
      [customerId],
    );

    if (rows.length === 0) {
      return { allowed: false, warning: false, used: 0, limit: 0, tier: "unknown" };
    }

    used = Number(rows[0].tokens_used);
    limit = Number(rows[0].tokens_limit);
    tier = String(rows[0].tier);

    await redis.set(
      cacheKey,
      JSON.stringify({ used, limit, tier }),
      "EX",
      LIMIT_CACHE_TTL,
    );
  }

  return {
    allowed: used < limit,
    warning: used >= Math.floor(limit * 0.9),
    used,
    limit,
    tier,
  };
}
